### ISSUE TITLE
Crash 2 Model Export

### ISSUE THREAD

**Issue by NetyashaRoozi** [cbhacks/CrashEdit#67]
I looked into a entity entity and am trying to rip ripper roo and it said obj but no save feature Idk if this is technically a issue for me or is not implemented yet ALSO I saw in screenshots that there is a model viewer but I cant find such viewer I can only view the scenery :\ I still love this program and hope someday there will be a update :)

**Comment by ManDude** [comment]
It's a Crash 1-only feature.

**Comment by NetyashaRoozi** [comment]
Thx ManDude hopefully somehow it will be implemented to crash 2-3 cause I wanted to get the crash 2  crash model and some levels

**Comment by warenhuis** [comment]
I don't know if this is related to the issue, but I am currently reverse-engineering the Crash 2 model format (the visual parts of it, at least).

I will write documentation about my findings soon but there are still some things that i need to work out. I might need some help because I dont have a lot of experience in game modding. Just wanted to let you know.

**Comment by ManDude** [comment]
Please do focus on the models (but I think someone on the CM Forums already decoded them but I never asked how). The **uncompressed** animation format for Crash 2/3 is already known.

**Comment by warenhuis** [comment]
If this page is the CM forum where you are reffering to, that was me:
http://hpzr.proboards.com/thread/6948/random-crash-thoughts?page=318
Otherwise I must've overlooked something, in which case I would like to receive a link.

Anyway, as you know there are really 2 types of models in Crash 2. The 5-item one and the 6-item one. I am almost done with understanding the 5-item one, but not the 6-item one (Crash Bandicoot's model is a 6-item one unfortunately).

For the 5-item modeltype I now have:
- scaling
- polygonlist
- colorlist
- uv map and textures

For the colors and location of the textures I need help from someone who knows how the texturechunks and palette-coloring work in memory. Some recommended reading would be nice.

The 6-item model is more difficult because it always corresponds to animations where the said amount of vertices in the T1-header (maybe these are morphtargets?) do not match up with the expected size of the item.

I'll wrap up the documentation for the 5-item one first.

**Comment by ManDude** [comment]
Oh, nice. Anyway, wurlitzerfox has detailed documentation as to how it maps textures right here: http://crash-hacking.wikia.com/wiki/World#Extracting_Texture_Information somewhere

I'm pretty sure each vertice maps to an index of the color list (unlike Crash 1, eugh),

Also, compressed frames have 3 extra Int32's in their header which might be compression settings (or weird indexes to the 6th item?).

**Comment by NetyashaRoozi** [comment]
Thank you for this amazing news! Now if only we could also get one for CTR so I can make custom tracks for games like MKWii and ect

**Comment by ughman** [comment]
Wow, I must have completely missed that post on the CM forums. This is really awesome.

There's some decent information on the playstation's GPU here:
http://problemkaputt.de/psx-spx.htm#gpurenderpolygoncommands
http://problemkaputt.de/psx-spx.htm#gpuvideomemoryvram

The important commands are probably these ones:

```
GP0(34h) - Shaded Textured three-point polygon, opaque, texture-blending
GP0(36h) - Shaded Textured three-point polygon, semi-transparent, tex-blend
GP0(3Ch) - Shaded Textured four-point polygon, opaque, texture-blending
GP0(3Eh) - Shaded Textured four-point polygon, semi-transparent, tex-blend
  1st  Color1+Command    (CcBbGgRrh)
  2nd  Vertex1           (YyyyXxxxh)
  3rd  Texcoord1+Palette (ClutYyXxh)
  4th  Color2            (00BbGgRrh)
  5th  Vertex2           (YyyyXxxxh)
  6th  Texcoord2+Texpage (PageYyXxh)
  7th  Color3            (00BbGgRrh)
  8th  Vertex3           (YyyyXxxxh)
  9th  Texcoord3         (0000YyXxh)
 (10th) Color4           (00BbGgRrh) (if any)
 (11th) Vertex4          (YyyyXxxxh) (if any)
 (12th) Texcoord4        (0000YyXxh) (if any)
```

From what I understand, `Page` selects which 256x256 section of VRAM will be used, and then Xx and Yy texcoords are within that area. `Clut * 16` would then be the address of the palette in VRAM, I suppose. I never actually managed to get texture rendering into any of CrashEdit's 3d viewers, so I'm not completely certain on any of this.

The person to talk to would be wurlitzerfox, but I'm not sure what the best way is to contact him/her.

There are some people on IRC in freenode #psxugnd and efnet #psxdev who may be more helpful.

**Comment by ughman** [comment]
Regarding the 6-item format, a few excerpts from Andy Gavin's online postings:

> We had a really sophisticated vertex packing algorithm to fit all the animation. It stored all the vertex positions using a 4 dimensional delta (both positional and temporal) using a bit based code so each vertex offset only used the minimum number of bits. It did around 30-50 to 1 compression!!! And the decompressor was all written in assembly.  
> http://all-things-andy-gavin.com/2011/02/04/making-crash-bandicoot-part-3/

&nbsp;

> The Crash characters have a crazy bit based compression scheme invented by Mark.  
> (not clear if he was talking about the platformers or CTR)
> http://all-things-andy-gavin.com/2012/04/29/diablo-3-commercial/

&nbsp;

> In addition, since Crash 2 has about 8 times the animation of the first game, we needed to really reduce the size of the data without sacrificing the quality of the animation.  After numerous rewrites the animation was stored as a special bitstream compressed in all 4 dimensions.  
> http://all-things-andy-gavin.com/2011/03/28/crash-bandicoot-teaching-an-old-dog-new-bits-part-3/

**Comment by NetyashaRoozi** [comment]
I think he was talking about crash platformer's after the first

**Comment by warenhuis** [comment]
Thanks for the tips everyone. I'll get into it as soon as possible. 

The 'vertex packing algorithm'  that mr. Gavin wrote is certainly familiar to me now. I will explain the algorithm soon.

The characters are indeed written differently, still have not figured it out yet.

The compression did very well for CB2, the game is about 4 times smaller in size than CB1.

**Comment by ManDude** [comment]
Well, Crash 1 has **lots** of obselete formats (mainly and mostly models).
- Each animation vertex uses 6 bytes (3 for normals and 3 for the actual coords).
- Model vertexes will repeat the same color and texture ref because there is no list to link to.
- Scenery doesn't support quads. Plus, it has the same issues as the model vertexes.

At least the zone format is very simple and small (and effective). Sadly the game will not function properly if you have over 400 objects in a single level. There's also the low polygon count.

Also, do note that Crash 1 uses dummy files. The actual size of (PAL) Crash 1 is a little over 200MB. NTSC-U Crash 3 is like 144MB which is the 2nd smallest in the entire franchise excluding handheld/mobile titles. If you're wondering what is the smallest, it's Bash at about 60MB.

**Comment by warenhuis** [comment]
My apologies, I deleted my post because I misspoke about the algorithm but you already responded.
But yeah, it is a strange development. 

The normals are obviously used for the character lighting in Crash 1, but in Crash 2 lighting doesn't really exist. All the artistic shadow relies solely on vertexcolors and darkness-flags.

The faces in Crash 2 do have 'directions', but they are indicated with a single half-byte.

**Comment by warenhuis** [comment]
Update: the vertexcoloring works almost as it is supposed to. There are still some out-of-place colors but they are rare. Every vertex in the animframe must be duplicated multiple times to support different vertexcolors, which is really tedious but necessary.
![models](https://cloud.githubusercontent.com/assets/15270225/12536794/21031d14-c2b0-11e5-8038-3264692a7e66.png)

I will delve into the texturing next. Thanks for the useful links by the way!
Together with the 6-item model decompressing, these will be the hardest parts by a long shot.

**Comment by NetyashaRoozi** [comment]
I cant wait till it gets ripped with textures, Soon we can start ripping characters and levels with full texture support, by soon I mean like in 2-10 weeks

**Comment by ManDude** [comment]
If it makes you feel any better, I [ripped some level models](https://sketchfab.com/mandudesmh).

**Comment by NetyashaRoozi** [comment]
Mandude that's amazing! are you going to attempt to put the textures on? does the Uv mapping data stay the same in the rip or does it change?

**Comment by ManDude** [comment]
Possibly? I don't think there is any format that supports vertex shading + texturing.

**Comment by warenhuis** [comment]
No, but PLY does support UV mapping - so it is just a matter of slamming on the texture once it is imported.

The challenge is that some UV segments use textures that are stretched (like 2 times as wide as they normally would be) while other segments use the normal dimensions.

**Comment by warenhuis** [comment]
![unt2](https://cloud.githubusercontent.com/assets/15270225/12584364/75126458-c447-11e5-92b8-430c89203f13.png)

Luckily, aside from the frameheaders, the CB3: Warped formats seem to be the same as the Crash 2 ones. Note how the groundshadows are just model geometry.

**Comment by ManDude** [comment]
I know about the shadows being textureless semi-transparent polygons (some shadows are also just textures). Not a bad idea actually.

**Comment by warenhuis** [comment]
Update: I made a working texture reader now. It is all in a experimental phase, but It won't be long before I release my documentation.

![untitled](https://cloud.githubusercontent.com/assets/15270225/12729778/a93d552c-c928-11e5-9b49-f72f52e01423.png)

**Comment by ughman** [comment]
Awesome.

I didn't even know they had an N on their uniforms.

**Comment by ManDude** [comment]
Please tell me that N was added by you because I literally never saw that.

**Comment by warenhuis** [comment]
There, the first fully textured CB2 export.
![image](https://cloud.githubusercontent.com/assets/15270225/12766438/21d7ef7a-ca04-11e5-9009-f73266f61a8a.png)
![image](https://cloud.githubusercontent.com/assets/15270225/12766509/8e33a538-ca04-11e5-9824-9143d1400c80.png)

Going to write the docs now.

**Comment by NetyashaRoozi** [comment]
Oooh the jet pack, I wanted to model that for my character but I could never find a good texture rip and my texturing skills are.... well not that good

**Comment by warenhuis** [comment]
Everything is generated automatically, so you don't have to worry about ripping.

**Comment by NetyashaRoozi** [comment]
thank you waren, do you have a program setup yet? also does it also work with ingame sprites?

**Comment by warenhuis** [comment]
Yes, I have a tool by which one can check:
- texturedata + UV map
- polygondata + vertexcolors

It also exports, but it only works with 5-item models as of yet.

What do you mean by ingame sprites?

**Comment by NetyashaRoozi** [comment]
Well I mean sprites, Idk why I put ingame in it

**Comment by ManDude** [comment]
If you mean stuff like the wumpa, you can't because its generated by code. You can just rip textures using PVV.

**Comment by warenhuis** [comment]
Are the HUD and the wumpa's 'models' or not?

**Comment by NetyashaRoozi** [comment]
No, They are not models BUT I have seen people rip them

**Comment by ghost** [comment]
I'm sure you could extract the flat planes that sprites are projected onto, relevant to the issue's topic.

**Comment by warenhuis** [comment]
If it is just sprites you're after, you can use software like 3D Ripper DX and an emulator. You can also read the texturechunks by loading them into PVV and find the right colorpalette that is included in the chunk.

**Comment by ManDude** [comment]
^Exactly. It's the only way to get a clean rip.

And no they're not models. The game generates quads using code ans maps textures to them.

**Comment by ghost** [comment]
Am I correct in thinking that this tool would also work for 5-item models in CB3?

**Comment by warenhuis** [comment]
Yes, as you can see with the medieval goat (in this issue), the tool only needs a few tweaks to get it to export CB3 models. I have not implemented it yet though.

I have to say, I've seen some terrible results with more complex models, so I'm not 100% sure.

**Comment by ghost** [comment]
Very good. I'm quite excited now.

**Comment by ManDude** [comment]
The models were made for low resolutions anyway (256x256x256 animations aren't exactly smooth).

**Comment by ughman** [comment]
> Are the HUD and the wumpa's 'models' or not?

I believe the GOOL entry 6th items define sprites, there's a writeup on how it works in Crash 1 here, but it does appear to match up with what I remember from Crash 2. http://crash-hacking.wikia.com/wiki/GOOL_Executable

Pretty sure the pause menu is a model, though.

**Comment by warenhuis** [comment]
https://github.com/warenhuis/Crash-Bandicoot-2-Modelexport

Ok I've put the tool online so that you can get your feet wet with the Crash 2 modeltypes. The documentation is also on there. The tool was written in the free edition of Gamemaker 7.0 (I'm not a real programmer).

The conversion is not 100% accurate yet as you will notice after some exports. Hopefully this will lead to new insights, because I will be busy for the coming months (I will follow a very intense semester at university).

**Comment by ManDude** [comment]
Dang, I've always wanted to write some sort of non-game program in GameMaker hehe.

Anyway this is good stuff! Maybe this can be ported and programmed into an actual programming language.

**Comment by ughman** [comment]
Awesome, I'm reading through the PDF now.

This vertex coloring scheme is surprisingly complex, a lot moreso than the scenery was/is. Guess I'll be rereading this several times over.

**Comment by warenhuis** [comment]
Glad you guys appreciate it. Imagine what the 6-item model has got in store. If I have any free time in the coming months, I will get back to it. We might need mr. Gavin on this project, lol.

**Comment by ughman** [comment]
Alright, I think I'm finally wrapping my head around this whole concept.

This format is sufficiently complicated that I don't think I can devise a single interface that every model format could "implement" to allow them all to share one renderer, one OBJ exporter, etc. Nor do I believe I can actually design an editor for this format, because this is truly insane. Modifying texture mappings does look relatively simple, though.

Instead, I will focus on developing an intermediate format. The various crash models would have code to convert to the intermediate format, and then the intermediate format would support rendering and exporting to various formats.

Will definitely have texture remapping functionality for the various models though, this is just too good.

**Comment by warenhuis** [comment]
Important quick update on the texturing:

In the docs I said that the 12-byte structure had the mapping points in order: 3, 2, 1.
It is more complicated however. The order depends on the "triangle type" that refers to the texinfo.
So for "triangle type "AA"" and "BB" this is still 3, 2, 1 but for "CC" it is 1, 2, 3 (the latter is probably true for scenery as well).

Then there is another texturetype which are the animated textures, that seem to follow another mapping system altogether (this is where the 5th item comes in I think), but it needs more research.

Edit: Got it fixed. I reuploaded the tool and documentation. To see what is fixed, please re-read item 3 and item 4 in the docs.

![untitled](https://cloud.githubusercontent.com/assets/15270225/12869375/b034af80-cd1d-11e5-953b-f7a439f579c7.png)

**Comment by NetyashaRoozi** [comment]
Oh wow I had no idea those things actually looked like that. Huh weird. Can I use the program to rip models for my game? Its free and its just a project for school

**Comment by ManDude** [comment]
Ohhh so that's what the red beam texture from the night ruins levels is for. I legit thought it was some unused thing all this time. The more you know...

Anyway WOW this is fucking brilliant. I am absolutely amazed.

**Comment by warenhuis** [comment]
Yes Harley, that is what the tool is for. I'm making a game of  my own as well. Looking forward to see what you do with it.

**Comment by ManDude** [comment]
So uh, since I _still_ don't quite get the format, what would I have to do to the model entry (BI10G) to remove the right side and back of an iron crate (for example)?

**Comment by warenhuis** [comment]
It is not easy to remove a face, because that will mess up the drawing order of the model. 

Consider the entire model like a 'tower' of blocks stacked on top of each other. Every triangle in the model is like a building block for the next and is essential to shape of the tower. 

What you can do, is change the textureinfo of some triangles to a fully transparent segment.
You're going to have to identify all the triangles (item 1 in items 0 to 4, or the second one if you will) on the box manually (for example, by changing the textureinfo or the colors and see the changes) and find the ones that you want to change. 

Every 4 byte-structure that is not a colorslot in item 1 defines a triangle (see my documentation for knowing when it is a colorslot or not). The first byte of such a structure refers to an index in item 3 (the textureinfo's, in 12 byte structures) ranging from 1 to 255, 0 = no texture.

If it is possible to change the size of entries in CrashEdit, eventually it will be possible to develop a model importer, making things much easier (like a ply to nsfmodel converter).

**Comment by ManDude** [comment]
CrashEdit has a "Replace Item" button (right-click). You cam add an "Add Item" button too.

Thanks though, but I want to actually remove the polygons because I wanted to save some polygons in an area to reduce lag.

**Comment by warenhuis** [comment]
Minor update: The first steps have been made to decompressing the animationvertices from the 6-item models. Below you can see a primitive example, the letter G from the titlescreen:
![untitled](https://cloud.githubusercontent.com/assets/15270225/14763734/cbe362cc-099e-11e6-84cd-311fd9d279a3.png) (The tiny yellow pixels are the vertices)

The '4 dimensional delta's' are stored within the 6th item of the model. The compressed animation frames seem to contain a bitstream of 3-bit structures (xzy). It is unclear as of yet how many different offsets they are capable of indicating (the first stream does an increment of 255).

The frame and model are interwoven in making the final shape of the model, but the 'temporal' dimension seems to do a lot of work in all 3 dimensions as well as in different frames. So that is where the next focus will be on.

**Comment by NetyashaRoozi** [comment]
Nice, we are so close to finally be able to rip all the models from Crash 2

**Comment by ManDude** [comment]
How high do you have to be to come up with such complex compression.

**Comment by warenhuis** [comment]
Minor update: Ok I think I now fully understand the compression format. Everything makes sense now, at least. If everything goes right, I'll have the first implemented rip soon.

**Comment by ManDude** [comment]
i lovyou

**Comment by warenhuis** [comment]
Ladies and gentlemen...
![untitled4](https://cloud.githubusercontent.com/assets/15270225/15044472/212c17ee-12d5-11e6-9670-f32e4950652b.png)
A fully clean decompression of Crash Bandicoot

Some more characters:
![untitled5](https://cloud.githubusercontent.com/assets/15270225/15046214/ea1782c0-12de-11e6-9ec5-3bb83038b6e5.png)

I'll still be working on some rough edges.

**Comment by ManDude** [comment]
Holy fucking SHIT. YOU DID IT. THIS IS BEYOND AWESOME.

I see you are using Blender... How are you displaying the textures and colors at the same time?

**Comment by warenhuis** [comment]
You can view the textures and colors in textureview, next to the menu for 'object mode', 'edit mode' etc..
Don't forget to remove lights from the scene + Uncheck mipmaps in User Preferences > System > Mipmaps to get accurate representation.

**Comment by NetyashaRoozi** [comment]
Ohmy..... This is amazing OMG! I AM SO HAPPY! We can finally get the best ripper roo!

**Comment by ughman** [comment]
Damn, guess I need to actually get to work.

**Comment by warenhuis** [comment]
https://github.com/warenhuis/Crash-Bandicoot-2-Modelexport

The tool now has 100% CB2 and CB3 model support. There are still some scaling issues between the two, I think (https://youtu.be/g7ByLSotvNo?t=2m10s). The proportions are correct though.

I've also written about the compression in 'item 5' and the 'frame' section of the documentation. Hopefully this will be enough for implementation in CrashEdit.

Enjoy!

**Comment by ManDude** [comment]
Have you thought about porting the project to GameMaker Studio?

I think the reason Crash 3 models are too small is because of the 0x30E property in entities, which determines scale (usually set to 1 in Crash 3, making objects 4x bigger).

**Comment by warenhuis** [comment]
So with these object entities, the final scale must be context-based? In that case this can only be solved in CrashEdit.

I can port to GameMaker Studio in an instance, but I don't see significant advantages of Studio over 7.

**Comment by ManDude** [comment]
Compatibility, maybe even speed? It lags a lot.

**Comment by warenhuis** [comment]
The tool is purely experimental. It lags mostly because of my horribly rushed coding style (there is a separate draw call for every number, even the off-screen ones). I can improve all that, but I think it would be much better if this gets implemented in CrashEdit right away.

**Comment by NetyashaRoozi** [comment]
![tiny](https://cloud.githubusercontent.com/assets/12688256/15104447/38431ada-157e-11e6-8245-9809ba2e947d.png)
I thought this guy was a 6 item guy ends up he isn't, Going to get a better frame with him since this one just looks awful

**Comment by NetyashaRoozi** [comment]
Btw, Tiny is under TZ for Taz tiger

**Comment by ManDude** [comment]
Ya we kinda know. ;)

**Comment by NetyashaRoozi** [comment]
I know, Just putting that information for anyone else trying to rip tiny

**Comment by warenhuis** [comment]
This is an important thing for scaling compatibility: Is there a way to know how big the wooden boxes are in absolute units? In CrashEdit they are 100 x 100 x 100 and 400 x 400 x 400 for CB2 and CB3 respectively, but the modeldata prescribes:
254 x 254 x 254, scaled x 1606 for CB2 and 
254 x 254 x 254, scaled x 201 for CB3?

How would this translate into the proper units?

**Comment by ManDude** [comment]
The scale changes a lot. In Crash 1 their animation is 64x64x64, their collision box is 102400x102400x102400 (roughly 102.4x102.4x102.4 in zone units), and they are 400x400x400 in-game (since zone units are in-game units divided by 4).

**Comment by supermariosunshine2002** [comment]
I really want to know how to export a good model from warped. can someone help me?

**Comment by AraHaan** [comment]
Dam I wish I had the source code to their model export for porting it to c++ and opengl or c# woth opengl. :thinking:

**Comment by ManDude** [comment]
soon
